File: src/main/java/tickets/BoardingPass.java
package tickets;


public class BoardingPass {
    public String boardingTime;
    public String gate;
    public String seat;
    public int ticketNum;
    public String passengerName;

    private BoardingPass(Builder builder) {
        this.boardingTime = builder.boardingTime;
        this.gate = builder.gate;
        this.seat = builder.seat;
        this.ticketNum = builder.ticketNum;
        this.passengerName = builder.passengerName;
    }


    public static class Builder {
        public String boardingTime;
        public String gate;
        public String seat;
        public int ticketNum;
        public String passengerName;

        public Builder(int ticketNum) {
            this.ticketNum = ticketNum;
        }

        public Builder isFlying(String passengerName) {
            this.passengerName = passengerName;
            return this;
        }

        public Builder withBoardingTime(String boardingTime) {
            this.boardingTime = boardingTime;
            return this;
        }

        public Builder atGate(String gate) {
            this.gate = gate;
            return this;
        }

        public Builder inSeat(String seat) {
            this.seat = seat;
            return this;
        }

        public BoardingPass build() {
            return new BoardingPass(this);
        }
    }
}

File: src/main/java/tickets/BoardingPass.kt
package tickets

import flights.Flight
import kotlin.random.Random

/**
 * Boarding pass Class. Includes the builder Design patterns as an internal class.
 */
class BoardingPass private constructor(builder: Builder) {
    var ticket: Ticket?
    var boardingTime: String?
    var gate: String?
    private var seat: String?
    var passenger: Passenger?

    init {
        boardingTime = builder.boardingTime
        gate = builder.gate
        seat = builder.seat
        ticket = builder.ticket
        passenger = builder.passenger
    }

    /**
     * Internal class Builder
     */
    class Builder {
        var boardingTime: String? = null
        var gate: String? = null
        var seat: String? = null
        lateinit var passenger: Passenger
        var ticket: Ticket? = null


        /**
         * link passenger to boarding pass
         */
        fun addPassenger(p: Passenger): Builder {
            this.passenger = p
            return this
        }


        fun addTicket(flight: Flight): Builder {
            flight.tp.acquireReusable()
            this.ticket = flight.tp.releaseTicket(passenger)
            return this
        }

        fun isFlying(): Builder {
            return this
        }

        /**
         * Checks API for boarding time.
         */
        fun withBoardingTime(): Builder {
            this.boardingTime = ticket?.flight?.expectedTimeGateOpen
            if (boardingTime == null) {
                println("Boarding Time not known yet")
            } else {
                println("Boarding time: $boardingTime")
            }
            return this
        }

        /**
         * Checks API for gate.
         */
        fun atGate(): Builder {
            this.gate = ticket?.flight?.gate
            if (gate == null) {
                println("Gate number not available yet")
            } else {
                println("Gate number: $gate")
            }
            return this
        }

        /**
         * gets a random seat on the flight
         * @exception e
         */
        fun inSeat(): Builder {
            return try {
                this.seat = ticket?.flight?.tp?.let { Random.nextInt(0, it.size).toString() }
                println("Seat number: $seat")
                this
            } catch (e: Exception) {
                e.printStackTrace()
                this
            }
        }

        /**
         * Actually builds the boarding pass and is always returned fully initialised
         * @return BoardingPass
         * @exception NullPointerException  check ticket is !null
         */
        fun build(): BoardingPass {
            if (this.ticket == null) {
                throw NullPointerException("There is no ticket assigned to this boardingpass")
            } else {
                return BoardingPass(this)
            }
        }
    }
}